flang/lib/Optimizer/OpenMP/DoConcurrentConversion.cpp

```diff
@@ -655,48 +663,101 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
   }
 
 private:
-  void genBoundsOps(mlir::ConversionPatternRewriter &rewriter,
-                    mlir::Location loc, hlfir::DeclareOp declareOp,
-                    llvm::SmallVectorImpl<mlir::Value> &boundsOps) const {
-    if (declareOp.getShape() == nullptr) {
+  struct TargetDeclareShapeCreationInfo {
+    std::vector<mlir::Value> startIndices{};
```

Comment:
Nit: I'd suggest adding a comment here explaining that we're using std::vector only to interface more easily with the `fir::ShapeShiftOp::getOrigins()`, etc methods, since it wouldn't otherwise be the preferred container to use.